File: Linked_list/java/linked_list/src/main/java/com/ayoola/linked_list/linked_list_implementation/LinkedList.java
package com.ayoola.linked_list.linked_list_implementation;

public class LinkedList {
    Node head;
    int number_of_nodes;

    public LinkedList(){
        this.head = null;
        this.number_of_nodes = 0;
    }

    @Override
    public String toString() {
        return "LinkedList with " + this.size_of_linked_list() + " number of nodes";
    }


    private Node create_node(int parameter){
        Node new_node;
        new_node = new Node(parameter);
        return new_node;
    }

    public int size_of_linked_list(){
        return this.number_of_nodes;
    }

    // traversing a linked list is an Ordo N O(N) running time complexity operation
    public void transverse() {
        Node actual_node = this.head;

        while (actual_node != null){
            System.out.println(actual_node.data);
            actual_node = actual_node.pointer;
        }
    }


    // constant running time complexity O(1)
    public void insert_start(int user_value){
        Node node = this.create_node(user_value);
        if (this.head != null) {
            node.pointer = this.head;
        }
        this.head = node;
        this.number_of_nodes++;
    }

    // linear running time complexity O(N)
    public void insert_end(int user_value){
        Node node = this.create_node(user_value);
        Node actual_node = this.head;

        while (actual_node.pointer != null){
            actual_node = actual_node.pointer;
        }
        actual_node.pointer = node;
        this.number_of_nodes++;
    }
}


class Node {
    int data;
    Node pointer;

    public Node(int value){
        this.data = value;
        this.pointer = null;
    }
}

File: Linked_list/python/linkedlist.py
class Node:

    def __init__(self, data):
        self.data = data
        self.next = None

    def __str__(self):
        try:
            return "data({0})||next({1})".format(self.data, self.next.data)
        except AttributeError:
            return "data({0})||next(None)".format(self.data)


class LinkedList:

    def __init__(self):
        self.root = None
        self.num_of_nodes = 0

    @staticmethod
    def create_node(data):
        node = Node(data)
        return node

    # Ordo 1 O(1) constant running time complexity
    def size_of_linked_list(self):
        return self.num_of_nodes

    # O(N) linear running time complexity
    def traverse(self):
        actual_node = self.root

        while actual_node is not None:
            print(actual_node.data)
            actual_node = actual_node.next

    # This has a O(N) running time complexity
    def search(self, data):
        actual_node = self.root

        while actual_node is not None and actual_node.data != data:
            actual_node = actual_node.next
        if actual_node is None:
            print("data not found")
        else:
            print("node:", actual_node)

    # constant running time O(1)
    def insert_start(self, data):
        node = self.create_node(data)
        if self.root is None:
            self.root = node
        else:
            node.next = self.root
            self.root = node
        self.num_of_nodes += 1

    # linear running time O(N)
    def insert_end(self, data):
        node = Node(data)
        actual_node = self.root

        while actual_node.next is not None:
            actual_node = actual_node.next
        else:
            actual_node.next = node
            self.num_of_nodes += 1

    def remove(self, data):
        node = self.create_node(data)
        # empty linked_list
        if self.root is None:
            return

        # Removing at the start - the root node. The running time complexity is O(1)
        if self.root.data == node.data:
            self.root = self.root.next
            self.num_of_nodes -= 1
        else:
            # Removing anywhere else in the linked_list. The running time complexity is O(N)
            actual_node = self.root
            previous_node = None
            while actual_node is not None and actual_node.data != node.data:
                previous_node = actual_node
                actual_node = actual_node.next

            # Value not in linked_list
            if actual_node is None:
                return
            # Value found within the linked list
            previous_node.next = actual_node.next
            self.num_of_nodes -= 1


if __name__ == "__main__":
    linked_list = LinkedList()
    linked_list.insert_start(23)
    linked_list.insert_start(10)
    linked_list.insert_start(3)
    linked_list.insert_end(100)
    linked_list.insert_end(150)
    print('There are ', linked_list.num_of_nodes, 'nodes in the linked list')
    linked_list.traverse()
    print("-----------------------------------------------------------------------")
    linked_list.remove(23)
    print('There are ', linked_list.num_of_nodes, 'nodes in the linked list')
    linked_list.traverse()
    print("-----------------------------------------------------------------------")
    linked_list.search(10)
